text/0000-unsafe-pointer-reform.md

```diff
@@ -0,0 +1,221 @@
+
+[The relevant FAQ entry](http://llvm.org/docs/GetElementPtr.html#what-happens-if-a-gep-computation-overflows) is as follows:
+
+> What happens if a GEP computation overflows?
+>
+> If the GEP lacks the inbounds keyword, the value is the result from evaluating the implied two’s complement integer computation. However, since there’s no guarantee of where an object will be allocated in the address space, such values have limited meaning.
+>
+> If the GEP has the inbounds keyword, the result value is undefined (a “trap value”) if the GEP overflows (i.e. wraps around the end of the address space).
+>
+> As such, there are some ramifications of this for inbounds GEPs: scales implied by array/vector/pointer indices are always known to be “nsw” since they are signed values that are scaled by the element size. These values are also allowed to be negative (e.g. “`gep i32 *%P, i32 -1`”) but the pointer itself is logically treated as an unsigned value. This means that GEPs have an asymmetric relation between the pointer base (which is treated as unsigned) and the offset applied to it (which is treated as signed). The result of the additions within the offset calculation cannot have signed overflow, but when applied to the base pointer, there can be signed overflow.
+
+This is written in a bit of a confusing way, so here's a simplified summary of what we care about: 
+
+* The pointer is treated as an unsigned number, and the offset as signed. 
+* While computing the offset in bytes (`idx * size_of::<T>()`), we aren't allowed to do signed overflow (nsw). 
+* While applying the offset to the pointer (`ptr + offset`), we aren't allowed to do unsigned overflow (nuw).
+
+Part of the historical argument for signed offset in Rust has been a *warning* against these overflow concerns, but upon inspection that doesn't really make sense. 
+
+* If you offset a `*const i16` by `isize::MAX / 3 * 2` (which fits into a signed integer), then you'll still overflow a signed integer in the implicit `offset` computation. 
+* There's no indication that unsigned overflow should be a concern at all.
+* The location of the offset *isn't even* the place to handle this issue. The ultimate consequence of `offset` being signed is that LLVM can't support allocations larger than `isize::MAX` bytes. Therefore this issue should be handled at the level of memory allocation code.
+* The fact that `offset` is `unsafe` is already surprising to anyone with the "it's just addition" mental model, pushing them to read the documentation and learn the actual rules.
+
+In conclusion: `as isize` doesn't actually help our developers write better code.
+
+
+
+
+# Detailed design
+[design]: #detailed-design
+
+
+## Methodization
+
+Add the following method equivalents for the static `ptr` functions on `*const T` and `*mut T`:
+
+```rust
+impl<T> *(const|mut) T {
```

Comment:
Will `*mut T` coerce to `*const T` sufficiently that we can impl these only on `*const T`?